apiserver/facades/agent/uniter/uniter.go

```diff
@@ -2466,25 +2486,108 @@ func (u *UniterAPI) GoalStates(args params.Entities) (string, error) {
 			result.Results[i].Error = common.ServerError(common.ErrPerm)
 			continue
 		}
-		// TODO (agprado): get units and relations information from
+		unit, err := u.getUnit(tag)
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		application, err := unit.Application()
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+		} else {
+			result.Results[i].Result, err = u.goalStateResult(application)
+			if err != nil {
+				result.Results[i].Error = common.ServerError(err)
+			}
+		}
+	}
+	return result, nil
+}
+
+// goalStateResult creates the structure with an application units and unit relations.
+func (u *UniterAPI) goalStateResult(application *state.Application) (*params.GoalState, error) {
+	gs := params.GoalState{}
+
+	var err error
+	gs.Units, err = u.goalStateUnits(application)
+	if err != nil {
+		return nil, err
+	}
+	allRelations, err := application.Relations()
+	if err != nil {
+		return nil, err
+	}
+	if allRelations != nil {
+		gs.Relations, err = u.goalStateRelations(allRelations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &gs, nil
+}
+
+// goalStateRelations creates the structure with all the relations between endpoints in an application.
+func (u *UniterAPI) goalStateRelations(allRelations []*state.Relation) (map[string]params.UnitsGoalState, error) {
 
-		result.Results[0].Result = "Hello World I'll be a yaml"
+	result := map[string]params.UnitsGoalState{}
+
+	logger.Debugf("receive %v relations", len(allRelations))
+	for _, r := range allRelations {
+
+		endPoints := r.Endpoints()
+		logger.Debugf("relation %v has %v endpoints", r.String(), len(endPoints))
+		for _, e := range endPoints {
+			if e.Relation.Role == "peer" {
+				continue
+			}
+			appName := e.ApplicationName
```

Comment:
Not sure of the value of this - we don't seem to use appName anywhere else so u.st.Application(e.ApplicationName) works fine I think